### DISCUSSION TITLE
caching in GenericBTRedutor

### DISCUSSION THREAD

**Discussion by artpelling** [pymor/pymor#1505] [Q&A] [answered]
Can somebody explain the reason why `GenericBTReductor` has an `_sv_U_V_cache` attribute instead of using the `@cached` decorator with its getter?

**Comment by sdrave** [comment, marked as answer]
Hi @artpelling, the reason is that the `cached` decorator is only available for immutable objects, but reductors are mutable (see #995). One possible fix would be to defer the caching to a small immutable class which only computes the gramians.

For some context: Reductors are mutable in pyMOR because  for RB methods the basis the reductor holds is iteratively extended by an outside greedy algorithm. From the RB prespective this is quite natural, but does not very well fit to the situation for most system-theoretic algorithms. Discussing the reductor design again (#1043) is on the agenda for after the 'great unification of the model hierarchy' @pmli and I are working on.

**Reply by artpelling** [reply]
Thank you! I had actually came across #995, but did not read it to the end :)
